### PR TITLE
Bindings with customizable DOM manipulators

### PR DESCRIPTION
**Pull request by Starzu**
Additional changes:
* bindings use `Node` instead of `Element`
* modification of listeners set inside listener works fine now

### CODE REVIEW

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
+package io.udash.bindings.modifiers
+
+import org.scalajs.dom.{Element, Node}
+
+private[bindings]
+trait DOMManipulator {
+  type ReplaceMethod = (Node, Seq[Node], Seq[Node]) => Boolean
+  type InsertMethod = (Node, Node, Seq[Node]) => Boolean
```
Comment (mkej) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: Those two should be moved to companion object.

File: core/frontend/src/main/scala/io/udash/properties/seq/SeqProperty.scala [hunk 2]
```diff
@@ -134,8 +134,11 @@ trait ReadableSeqProperty[A, +ElemType <: ReadableProperty[A]] extends ReadableP
   def nonEmpty: Boolean =
     elemProperties.nonEmpty
 
-  protected def fireElementsListeners[ItemType <: ReadableProperty[A]](patch: Patch[ItemType], structureListeners: Iterable[(Patch[ItemType]) => Any]): Unit =
-    CallbackSequencer.queue(s"${this.id.toString}:fireElementsListeners:${patch.hashCode()}", () => structureListeners.foreach(_.apply(patch)))
+  protected def fireElementsListeners[ItemType <: ReadableProperty[A]](patch: Patch[ItemType], structureListeners: Iterable[(Patch[ItemType]) => Any]): Unit = {
+    val cpy = mutable.Set.empty[(Patch[ItemType]) => Any]
```
Comment (mkej) [inline core/frontend/src/main/scala/io/udash/properties/seq/SeqProperty.scala]:
```scala
val cpy = structureListeners.toSet
```

File: core/frontend/src/main/scala/io/udash/properties/single/Property.scala [hunk 3]
```diff
@@ -151,7 +151,9 @@ trait ReadableProperty[A] {
   protected[properties] def fireValueListeners(): Unit = {
     CallbackSequencer.queue(s"${this.id.toString}:fireValueListeners", () => {
       val t = get
-      listeners.foreach(_.apply(t))
+      val cpy = mutable.Set.empty[A => Any]
```
Comment (mkej) [inline core/frontend/src/main/scala/io/udash/properties/single/Property.scala]:
```scala
val cpy = listeners.toSet
```

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package io.udash.bindings.modifiers
+
+import org.scalajs.dom.{Element, Node}
+
+private[bindings]
+trait DOMManipulator {
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: What's this line break for?

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala (hunk 1 above)
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: I cannot see a reason for using type members here. The types are not overridden anywhere, I'm not sure they even should be and if these contracts were just a part of the method signatures in this trait, the arguments would have meaningful names.

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/package.scala [hunk 5]
```diff
@@ -1,10 +1,10 @@
 package io.udash.bindings
 
-import org.scalajs.dom.Element
+import org.scalajs.dom.Node
 
 package object modifiers {
-  implicit class ElementExts(val el: Element) extends AnyVal {
-    def replaceChildren(oldChildren: Seq[Element], newChildren: Seq[Element]): Unit = {
+  implicit class ElementExts(val el: Node) extends AnyVal {
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/package.scala]: private val

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala (hunk 4 above)
Comment (Starzu) [reply core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: Consistency with the other classes in this package.

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  import DOMManipulator._
+
+  /**
+    * Provides custom child elements replace method. This method takes
+    * root element, old children and new children. It should return `true`,
+    * if it does not replace elements in DOM.
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: The result requirements should be explained, so the user understands what that boolean triggers.

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    */
+  def customElementsReplace: ReplaceMethod
+
+  /**
+    * Provides custom child elements insert method. This method takes
+    * root element, ref node and new children. It should return `true`,
+    * if it does not insert elements in DOM.
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: Same here.

File: core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      newElements.foreach(root.insertBefore(_, before))
+    }
+}
+
+private[bindings]
+object DOMManipulator {
+  type ReplaceMethod = (Node, Seq[Node], Seq[Node]) => Boolean
+  type InsertMethod = (Node, Node, Seq[Node]) => Boolean
+
+  val defaultElementReplace: ReplaceMethod = (_, _, _) => true
```
Comment (ddworak) [inline core/frontend/src/main/scala/io/udash/bindings/modifiers/DOMManipulator.scala]: Make these final and use upper-camel case for constants.

**Pull request merged**